sct-commons/src/main/java/org/lfenergy/compas/sct/commons/DataTypeTemplatesService.java

```diff
@@ -33,14 +42,87 @@ public class DataTypeTemplatesService {
      */
     public boolean isDoModAndDaStValExist(TDataTypeTemplates dtt, String lNodeTypeId) {
         return lnodeTypeService.findLnodeType(dtt, lNodeType -> lNodeTypeId.equals(lNodeType.getId()))
-                .map(lNodeType -> doService.findDo(lNodeType, tdo -> MOD_DO_NAME.equals(tdo.getName()))
-                        .map(tdo -> doTypeService.findDoType(dtt, doType -> tdo.getType().equals(doType.getId()))
-                                .map(doType -> doType.getSDOOrDA().stream()
-                                                .filter(sdoOrDa -> sdoOrDa.getClass().equals(TDA.class))
-                                                .map(TDA.class::cast)
-                                                .anyMatch(tda -> STVAL_DA_NAME.equals(tda.getName())))
-                                 .orElse(false))
-                         .orElse(false))
-                 .orElse(false);
+                .flatMap(lNodeType -> doService.findDo(lNodeType, tdo -> MOD_DO_NAME.equals(tdo.getName()))
+                        .flatMap(tdo -> doTypeService.findDoType(dtt, doType -> tdo.getType().equals(doType.getId()))
+                                .map(doType -> sdoOrDAService.findDA(doType, tda -> STVAL_DA_NAME.equals(tda.getName())).isPresent())))
+                .orElse(false);
+    }
+
+    @Override
+    public Stream<DataAttributeRef> getAllDOAndDA(TDataTypeTemplates dtt) {
+        return dtt.getLNodeType().stream()
+                .flatMap(tlNodeType -> lnodeTypeService.getAllDOAndDA(dtt, tlNodeType.getId()));
```

Comment:
remove the `lnodeTypeService.getAllDOAndDA` method; it does not belong to LnodeTypeService; 
and use steps in your lambda to know where you are
```java
public Stream<DataAttributeRef> getAllDOAndDA(TDataTypeTemplates dtt) {
        return lnodeTypeService.getLnodeTypes(dtt)
                .flatMap(tlNodeType -> {
                    DataAttributeRef dataAttributeRef = new DataAttributeRef();
                    dataAttributeRef.setLnType(tlNodeType.getId());
                    if (tlNodeType.isSetLnClass() && !tlNodeType.getLnClass().isEmpty()) {
                        dataAttributeRef.setLnClass(tlNodeType.getLnClass().getFirst());
                    }
                    return tlNodeType.getDO()
                            .stream()
                            .map(tdo -> {
                                dataAttributeRef.getDoName().setName(tdo.getName());
                                return doTypeService.findDoType(dtt, tdoType -> tdoType.getId().equals(tdo.getType()))
                                        .map(doType -> {
                                            dataAttributeRef.getDoName().setCdc(doType.getCdc());
                                            return doTypeService.getAllSDOAndDA(dtt, doType, dataAttributeRef).stream();
                                        });
                            })
                            .filter(Optional::isPresent)
                            .flatMap(Optional::orElseThrow);
                });
    }
```